File: data/src/main/java/com/microsoft/azure/kusto/data/auth/endpoints/WellKnownKustoEndpointsData.java
package com.microsoft.azure.kusto.data.auth.endpoints;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.microsoft.azure.kusto.data.http.HttpPostUtils;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.HashMap;

public class WellKnownKustoEndpointsData {
    public static class AllowedEndpoints {
        public ArrayList<String> AllowedKustoSuffixes;
        public ArrayList<String> AllowedKustoHostnames;
    }

    @JsonIgnore
    public ArrayList<String> _Comments;
    public HashMap<String, AllowedEndpoints> AllowedEndpointsByLogin;
    private static WellKnownKustoEndpointsData instance = null;
    private static final Object object = new Object();

    public static WellKnownKustoEndpointsData getInstance() {
        if (instance != null)
            return instance;
        synchronized (object) {
            if (instance == null) {
                instance = readInstance();
            }
        }
        return instance;
    }

    // For Deserialization
    public WellKnownKustoEndpointsData() {
    }

    private static WellKnownKustoEndpointsData readInstance() {
        try {
            // Beautiful !
            ObjectMapper objectMapper = HttpPostUtils.getObjectMapper();
            try (InputStream resourceAsStream = WellKnownKustoEndpointsData.class.getResourceAsStream(
                    "/WellKnownKustoEndpoints.json")) {
                return objectMapper.readValue(resourceAsStream, WellKnownKustoEndpointsData.class);
            }
        } catch (Exception ex) {
            ex.printStackTrace();
        }
        return null;
    }
}

File: ingest-v2/src/main/kotlin/com/microsoft/azure/kusto/ingest/v2/auth/endpoints/WellKnownKustoEndpointsData.kt
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
package com.microsoft.azure.kusto.ingest.v2.auth.endpoints

import kotlinx.serialization.SerialName
import kotlinx.serialization.json.Json
import org.slf4j.LoggerFactory
import kotlinx.serialization.Serializable as KSerializable

/** Data class representing the structure of WellKnownKustoEndpoints.json */
@KSerializable
data class AllowedEndpoints(
    @SerialName("AllowedKustoSuffixes")
    val allowedKustoSuffixes: List<String> = emptyList(),
    @SerialName("AllowedKustoHostnames")
    val allowedKustoHostnames: List<String> = emptyList(),
)

@KSerializable
data class WellKnownKustoEndpointsData(
    @SerialName("_Comments") val comments: List<String> = emptyList(),
    @SerialName("AllowedEndpointsByLogin")
    val allowedEndpointsByLogin: Map<String, AllowedEndpoints> = emptyMap(),
) {
    companion object {
        private val logger =
            LoggerFactory.getLogger(WellKnownKustoEndpointsData::class.java)

        @Volatile private var instance: WellKnownKustoEndpointsData? = null

        private val json = Json {
            ignoreUnknownKeys = true
            isLenient = true
        }

        fun getInstance(): WellKnownKustoEndpointsData {
            return instance
                ?: synchronized(this) {
                    instance ?: readInstance().also { instance = it }
                }
        }

        private fun readInstance(): WellKnownKustoEndpointsData {
            return try {
                val resourceStream =
                    WellKnownKustoEndpointsData::class
                        .java
                        .getResourceAsStream(
                            "/WellKnownKustoEndpoints.json",
                        )
                        ?: throw RuntimeException(
                            "WellKnownKustoEndpoints.json not found in classpath",
                        )

                val content =
                    resourceStream.bufferedReader().use { it.readText() }
                json.decodeFromString<WellKnownKustoEndpointsData>(content)
            } catch (ex: Exception) {
                logger.error("Failed to read WellKnownKustoEndpoints.json", ex)
                throw RuntimeException(
                    "Failed to read WellKnownKustoEndpoints.json",
                    ex,
                )
            }
        }
    }
}
